presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,762 @@
+
+        if (bytes.length > UNSCALED_DECIMAL_128_SLICE_LENGTH
+                || (bytes.length == UNSCALED_DECIMAL_128_SLICE_LENGTH && (bytes[0] & SIGN_BYTE_MASK) != 0)) {
+            throwOverflowException();
+        }
+
+        // convert to little-endian order
+        reverse(bytes);
+
+        Slice decimal = Slices.allocate(UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        decimal.setBytes(0, bytes);
+        if (unscaledValue.signum() < 0) {
+            setNegative(decimal, true);
+        }
+
+        throwIfOverflows(decimal);
+
+        return decimal;
+    }
+
+    public static Slice unscaledDecimal(long unscaledValue)
+    {
+        long[] longs = new long[NUMBER_OF_LONGS];
+        if (unscaledValue < 0) {
+            longs[0] = -unscaledValue;
+            longs[1] = SIGN_LONG_MASK;
+        }
+        else {
+            longs[0] = unscaledValue;
+        }
+        return Slices.wrappedLongArray(longs);
+    }
+
+    public static BigInteger unscaledDecimalToBigInteger(Slice decimal)
+    {
+        byte[] bytes = decimal.getBytes(0, UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        // convert to big-endian order
+        reverse(bytes);
+        bytes[0] &= ~SIGN_BYTE_MASK;
+        return new BigInteger(isNegative(decimal) ? -1 : 1, bytes);
+    }
+
+    public static long unscaledDecimalToUnscaledLong(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        long hi = getLong(decimal, 1);
+        boolean negative = isNegative(decimal);
+
+        if (hi != 0 || ((lo > Long.MIN_VALUE || !negative) && lo < 0)) {
+            throwOverflowException();
+        }
+
+        return negative ? -lo : lo;
+    }
+
+    public static long unscaledDecimalToUnscaledLongUnsafe(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        return isNegative(decimal) ? -lo : lo;
+    }
+
+    public static Slice rescale(Slice decimal, int rescaleFactor)
+    {
+        if (rescaleFactor == 0) {
+            return decimal;
+        }
+        else {
+            Slice result = unscaledDecimal();
+            rescale(decimal, rescaleFactor, result);
+            return result;
+        }
+    }
+
+    public static void rescale(Slice decimal, int rescaleFactor, Slice result)
+    {
+        if (rescaleFactor == 0) {
+            copyUnscaledDecimal(decimal, result);
+        }
+        else if (rescaleFactor > 0) {
+            if (rescaleFactor >= POWERS_OF_TEN.length) {
+                throwOverflowException();
+            }
+            multiply(decimal, POWERS_OF_TEN[rescaleFactor], result);
+        }
+        else {
+            scaleDownRoundUp(decimal, -rescaleFactor, result);
+        }
+    }
+
+    public static Slice multiply(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+        multiply(left, right, result);
+        return result;
+    }
+
+    public static void multiply(Slice left, Slice right, Slice result)
+    {
+        int l0 = getInt(left, 0);
```

Comment:
What benchmark did you use?

I have changed the code to:
```
        long l0 = getInt(left, 0) & LONG_MASK;
        long l1 = getInt(left, 1) & LONG_MASK;
        long l2 = getInt(left, 2) & LONG_MASK;
        long l3 = getInt(left, 3) & LONG_MASK;

        long r0 = getInt(right, 0) & LONG_MASK;
        long r1 = getInt(right, 1) & LONG_MASK;
        long r2 = getInt(right, 2) & LONG_MASK;
        long r3 = getInt(right, 3) & LONG_MASK;
```
In order to avoid direct manipulation on llow-level `Slice`. JIT should inline `getInt` method call anyway I think.

My benchmark results:
```
Old:

Benchmark                                    (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.multiplyBenchmark       l1 * l2  thrpt   30  0,404 ± 0,012  ops/ms

My:

Benchmark                                    (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.multiplyBenchmark       l1 * l2  thrpt   30  0,424 ± 0,017  ops/ms

Martin:

Benchmark                                    (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.multiplyBenchmark       l1 * l2  thrpt   30  0,423 ± 0,011  ops/ms
```